### PR TITLE
GetNextList func added, fix for tests

### PR DESCRIPTION
**Pull request by aav66**

### CODE REVIEW

File: s3_test.go [hunk 1]
```diff
@@ -87,9 +85,10 @@ func TestExportImport(t *testing.T) {
 	}
 	defer os.RemoveAll("./test_data")
 
-	if s, err = testInit(); err != nil {
-		t.Fatal(err)
-	}
+	//	if s, err = testInit(); err != nil {
```
Comment (itsmontoya) [inline s3_test.go:88]: If we don't need this anymore, can we remove?

File: s3_test.go [hunk 2]
```diff
@@ -121,19 +120,30 @@ func TestExportImport(t *testing.T) {
...
+	//o.AvoidBucketCreation = true
+	o.Endpoint = "https://nyc3.digitaloceanspaces.com"
 	return New(o)
 }
 
 func testClose(t *testing.T, s *S3) {
-	if err := s.deleteBucket(); err != nil {
-		t.Fatalf("Error encountered while deleting: %v\n", err)
-	}
+	//if err := s.deleteBucket(); err != nil {
```
Comment (itsmontoya) [inline s3_test.go:146]: Did you mean to leave this out?

**Review by itsmontoya — CHANGES_REQUESTED** [review]

File: s3.go [hunk 3]
```diff
@@ -170,6 +170,36 @@ func (s *S3) GetNext(ctx context.Context, prefix, lastFilename string) (nextKey
...
+	input := s3.ListObjectsV2Input{
+		Bucket:     aws.String(s.o.Bucket),
+		Prefix:     aws.String(prefix + "/"),
+		StartAfter: aws.String(startAfter),
+		MaxKeys:    aws.Int64(maxkeys),
+	}
+
+	var out *s3.ListObjectsV2Output
+	if out, err = s.s3.ListObjectsV2WithContext(ctx, &input); err != nil {
+		s.getNextErrored.With(prometheus.Labels{"aws_error": getAwsErrorCode(err)}).Inc()
```
Comment (itsmontoya) [inline s3.go:185]: I'd rather do hooks. So provide funcs in the opts for this. That way we aren't introducing dependencies.

So in the options do OnGetNextError, as an example

File: s3.go [hunk 4]
```diff
@@ -170,6 +170,36 @@ func (s *S3) GetNext(ctx context.Context, prefix, lastFilename string) (nextKey
...
+		return
+	}
+
+	if len(out.Contents) == 0 {
+		err = io.EOF
+		return
+	}
+
+	nextKeys = make([]string, len(out.Contents))
```
Comment (itsmontoya) [inline s3.go:194]: Even though your approach is technically more performance. The Go best practices leans on safety instead. Safer to make with capacity and grow using append:

```go
nextKeys = make([]string, 0, len(out.Contents))
for i, v := range out.Contents {
	nextKeys = append(nextKeys, path.Base(*v.Key))
}

```

**Review by itsmontoya — CHANGES_REQUESTED** [review]
A few more change requests!

File: s3.go (hunk 4 above)
Comment (aav66) [reply s3.go:194]: I think we safe.... we know a number of files.

Comment (aav66) [reply s3.go:194]: but yep... I will update it

Comment (itsmontoya) [reply s3.go:194]: > I think we safe.... we know a number of files.

Yes I agree, it's definitely safe. This is just about producing idiomatic Go

File: s3_test.go (hunk 1 above)
Comment (aav66) [reply s3_test.go]: decide to keep it for cleanup

File: s3.go (hunk 4 above)
Comment (aav66) [reply s3.go]: fixed

**Review by itsmontoya — APPROVED** [review]

**Pull request merged**